File: tools/convert_tests/scenarios/js_to_strict_ts_upgrade/behavior_check.js
#!/usr/bin/env node
/**
 * Behavior preservation check for JS to Strict TS conversion.
 * This script tests that key behavior remains unchanged after conversion.
 */

const { spawn } = require('child_process');
const crypto = require('crypto');

function runJavaScriptCode(filePath) {
    return new Promise((resolve, reject) => {
        const child = spawn('node', [filePath], {
            timeout: 30000,
            stdio: ['pipe', 'pipe', 'pipe']
        });

        let stdout = '';
        let stderr = '';

        child.stdout.on('data', (data) => {
            stdout += data.toString();
        });

        child.stderr.on('data', (data) => {
            stderr += data.toString();
        });

        child.on('close', (code) => {
            resolve({ stdout, stderr, returnCode: code });
        });

        child.on('error', (error) => {
            reject(error);
        });
    });
}

function calculateOutputHash(output) {
    return crypto.createHash('md5').update(output).digest('hex');
}

async function testJSBehavior() {
    console.log('Testing JavaScript behavior preservation...');

    try {
        // Run the calculator module from source_repo
        const result = await runJavaScriptCode('source_repo/calculator.js');
        
        if (result.returnCode !== 0) {
            console.error(`ERROR: JavaScript code execution failed with return code ${result.returnCode}`);
            console.error(`STDERR: ${result.stderr}`);
            return false;
        }

        // Check for expected outputs in the calculator execution
        const expectedOutputs = [
            'Initial value: 10',
            'After adding 5: 15', 
            'After subtracting 3: 12',
            'Sum of 5 and 3: 8',
            'Multiply 4 and 5: 20'
        ];
        
        const missingOutputs = [];
        expectedOutputs.forEach(expected => {
            if (!result.stdout.includes(expected)) {
                missingOutputs.push(expected);
            }
        });

        if (missingOutputs.length > 0) {
            console.error(`ERROR: Missing expected outputs:`, missingOutputs);
            console.error(`Actual output:`, result.stdout);
            return false;
        }

        // Calculate hash of the output to compare against baseline
        const outputHash = calculateOutputHash(result.stdout);
        console.log(`Output hash: ${outputHash}`);
        
        console.log('JavaScript behavior preservation test PASSED');
        return true;
    } catch (error) {
        console.error('ERROR:', error.message);
        return false;
    }
}

// Run the test
testJSBehavior().then(success => {
    process.exit(success ? 0 : 1);
});

File: tools/convert_tests/scenarios/python_to_typed_python_basic/behavior_check.py
#!/usr/bin/env python3
"""
Behavior preservation check for Python to Typed Python conversion.
This script tests that key behavior remains unchanged after conversion.
"""

import subprocess
import sys
import os
import hashlib

def run_python_code(file_path):
    """Run the Python code and capture output."""
    try:
        result = subprocess.run([sys.executable, file_path], 
                              capture_output=True, text=True, timeout=30)
        return result.stdout, result.stderr, result.returncode
    except subprocess.TimeoutExpired:
        return "", "Timeout occurred", 1

def calculate_output_hash(output):
    """Calculate hash of output to compare behavior."""
    return hashlib.md5(output.encode()).hexdigest()

def test_python_behavior():
    """Test that Python code behavior is preserved."""
    print("Testing Python behavior preservation...")

    # Run the calculator module from source_repo
    stdout, stderr, returncode = run_python_code("source_repo/calculator.py")
    
    if returncode != 0:
        print(f"ERROR: Python code execution failed with return code {returncode}")
        print(f"STDERR: {stderr}")
        return False
    
    # Check for expected outputs in the calculator execution
    expected_outputs = [
        "Initial value: 10",
        "After adding 5: 15",
        "After subtracting 3: 12",
        "Sum of 5 and 3: 8",
        "Multiply 4 and 5: 20"
    ]
    
    missing_outputs = []
    for expected in expected_outputs:
        if expected not in stdout:
            missing_outputs.append(expected)
    
    if missing_outputs:
        print(f"ERROR: Missing expected outputs: {missing_outputs}")
        print(f"Actual output: {stdout}")
        return False
    
    # Calculate hash of the output to compare against baseline
    output_hash = calculate_output_hash(stdout)
    print(f"Output hash: {output_hash}")
    
    # For the preservation check, we just ensure the expected outputs are present
    print("Python behavior preservation test PASSED")
    return True

if __name__ == "__main__":
    success = test_python_behavior()
    sys.exit(0 if success else 1)